File: pypp_cli/test_dir/cpp/src/exceptions/throw_.cpp
#include "exceptions/throw_.h"
#include "exceptions/exception.h"
#include "exceptions/stdexcept.h"
#include "py_str.h"
#include "pypp_util/print.h"
#include "pypp_util/to_py_str.h"
#include <string>

namespace me {
void _test_bare_raise() {
    try {
        throw pypp::TypeError(pypp::PyStr("test"));
    } catch (const pypp::TypeError &) {
        throw;
    }
}

void throw_fn() {
    pypp::print(pypp::PyStr("EXCEPTION RESULTS:"));
    try {
        throw pypp::Exception(pypp::PyStr("test"));
    } catch (const pypp::Exception &) {
        pypp::print(pypp::PyStr("exception happened"));
    }
    try {
        throw pypp::TypeError(pypp::PyStr("test"));
    } catch (const pypp::TypeError &) {
        pypp::print(pypp::PyStr("type error caught"));
    }
    try {
        throw pypp::TypeError(pypp::PyStr("test"));
    } catch (const pypp::TypeError &pypp_e) {
        std::string e = pypp_e.what();
        pypp::print(pypp::PyStr("type error caught: ") + pypp::str(e));
    }
    try {
        throw pypp::TypeError(pypp::PyStr("test"));
    } catch (const pypp::TypeError &) {
        pypp::print(pypp::PyStr("type error caught"));
    } catch (const pypp::ValueError &) {
        pypp::print(pypp::PyStr("value error caught"));
    } catch (const pypp::Exception &) {
        pypp::print(pypp::PyStr("other error caught"));
    }
    try {
        throw pypp::TypeError(pypp::PyStr("test"));
    } catch (...) {
        pypp::print(pypp::PyStr("catching all"));
    }
    try {
        _test_bare_raise();
    } catch (const pypp::TypeError &) {
        pypp::print(pypp::PyStr("bare raise"));
    }
}

} // namespace me

File: pypp_cli/test_dir/python/src/exceptions/throw_.py
def _test_bare_raise():
    try:
        raise TypeError("test")
    except TypeError:
        raise


def throw_fn():
    print("EXCEPTION RESULTS:")
    # raising an exception and catching all
    try:
        raise Exception("test")
    except Exception:
        print("exception happened")
    # raise an exception and catch specific
    try:
        raise TypeError("test")
    except TypeError:
        print("type error caught")
    # raise an exception and catch specific with the catch value
    try:
        raise TypeError("test")
    except TypeError as e:
        print("type error caught: " + str(e))
    # catch multiple types
    try:
        raise TypeError("test")
    except TypeError:
        print("type error caught")
    except ValueError:
        print("value error caught")
    except Exception:
        print("other error caught")
    # all in one
    try:
        raise TypeError("test")
    except:
        print("catching all")

    try:
        _test_bare_raise()
    except TypeError:
        print("bare raise")
